addons/iap/models/iap_account.py

```diff
@@ -74,32 +161,49 @@ def get(self, service_name, force_create=True):
         return accounts[0]
 
     @api.model
-    def get_credits_url(self, service_name, base_url='', credit=0, trial=False):
+    def get_credits_url(self, service_name, base_url='', credit=0, trial=False, account_token=False, get_service_from_token=False):
         """ Called notably by ajax crash manager, buy more widget, partner_autocomplete, sanilmail. """
         dbuuid = self.env['ir.config_parameter'].sudo().get_param('database.uuid')
         if not base_url:
             endpoint = iap_tools.iap_get_endpoint(self.env)
             route = '/iap/1/credit'
             base_url = endpoint + route
-        account_token = self.get(service_name).account_token
+        if not account_token:
+            account_token = self.get(service_name).account_token
         d = {
             'dbuuid': dbuuid,
             'service_name': service_name,
             'account_token': account_token,
             'credit': credit,
+            'get_service_from_token': get_service_from_token,
         }
         if trial:
             d.update({'trial': trial})
         return '%s?%s' % (base_url, werkzeug.urls.url_encode(d))
 
+    def action_buy_credits(self):
+        for account in self:
+            return {
+                'type': 'ir.actions.act_url',
+                'url': self.env['iap.account'].get_credits_url(
+                    account_token=account.account_token,
+                    service_name=account.service_name,
+                    get_service_from_token=True,
+                ),
+            }
+
+    def action_toggle_show_token(self):
+        for account in self:
+            account.show_token = not account.show_token
     @api.model
     def get_account_url(self):
         """ Called only by res settings """
         route = '/iap/services'
         endpoint = iap_tools.iap_get_endpoint(self.env)
-        d = {'dbuuid': self.env['ir.config_parameter'].sudo().get_param('database.uuid')}
-
-        return '%s?%s' % (endpoint + route, werkzeug.urls.url_encode(d))
+        data = {
+            'dbuuid': self.env['ir.config_parameter'].sudo().get_param('database.uuid'),
+        }
+        return '%s?%s' % (endpoint + route, werkzeug.urls.url_encode(data))
```

Comment:
I know that I've asked you to change the `d` variable into something more explicit, but why was there a diff here in the first place ?